File: Java/calculator/TollCalculator.java
package calculator;

import util.Day;
import util.Precondition;

import java.util.Calendar;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.concurrent.TimeUnit;

public class TollCalculator {

    private final Specifications specifications;

    public TollCalculator(Specifications specifications)
    {
        Precondition.isNotNull(specifications, "specifications");

        this.specifications = specifications;
    }

    public TollCalculator()
    {
        this(Specifications.newDefault());
    }

    /**
     * Calculate the total toll fee for one day
     *
     * @param vehicle - the vehicle
     * @param dates   - date and time of all passes on one day
     * @return - the total toll fee for that day
     */
    public int getTollFee(Vehicle vehicle, Date... dates)
    {
        Date intervalStart = dates[0];
        int totalFee = 0;
        for (Date date : dates) {
            int nextFee = getTollFee(date, vehicle);
            int tempFee = getTollFee(intervalStart, vehicle);

            TimeUnit timeUnit = TimeUnit.MINUTES;
            long diffInMillies = date.getTime() - intervalStart.getTime();
            long minutes = timeUnit.convert(diffInMillies, TimeUnit.MILLISECONDS);

            if (minutes <= 60) {
                if (totalFee > 0) totalFee -= tempFee;
                if (nextFee >= tempFee) tempFee = nextFee;
                totalFee += tempFee;
            } else {
                totalFee += nextFee;
            }
        }
        if (totalFee > 60) totalFee = 60;
        return totalFee;
    }

    public int getTollFee(final Date date, Vehicle vehicle)
    {
        if (isTollFreeVehicle(vehicle)) return 0;

        Calendar dateTime = dateTimeOf(date);

        if (isTollFreeDate(dateTime)) return 0;

        return getFeeForTimeOfDay(dateTime);
    }

    private int getFeeForTimeOfDay(Calendar dateTime)
    {
        int hour = dateTime.get(Calendar.HOUR_OF_DAY);
        int minute = dateTime.get(Calendar.MINUTE);
        return specifications.feeForTimeOfDay.feeFor(hour, minute);
    }

    private boolean isTollFreeVehicle(Vehicle vehicle)
    {
        return specifications.isTollFreeVehicle.test(vehicle);
    }

    private boolean isTollFreeDate(Calendar dateTime)
    {
        return isWeekend(dateTime) || specifications.isHoliday.test(new Day(dateTime));
    }

    private boolean isWeekend(Calendar dateTime)
    {
        int dayOfWeek = dateTime.get(Calendar.DAY_OF_WEEK);
        return dayOfWeek == Calendar.SATURDAY || dayOfWeek == Calendar.SUNDAY;
    }

    private static Calendar dateTimeOf(Date date)
    {
        Calendar calendar = GregorianCalendar.getInstance();
        calendar.setTime(date);
        return calendar;
    }

}

File: C#/TollCalculationApp/TollCalculator.cs
using System;
using System.Collections.Generic;
using TollFeeCalculator;
using Nager.Date;

public static class TollCalculator
{

    /**
     * Calculate the total toll fee for one day
     *
     * @param vehicle - the vehicle
     * @param dates   - date and time of all passes on one day
     * @return - the total toll fee for that day
     * 
     * Note that Nager.Date runs under MIT license!
     */

    public static int GetTollFee(IVehicle vehicle, DateTime[] dates)
    {
        DateTime intervalStart = dates[0];
        List<int> fees = new List<int>();
        foreach( DateTime pass in dates)
        {
            fees.Add(GetTollFee(pass, vehicle));
        }
        int totalFee = 0;
        foreach (DateTime date in dates)
        {
            int nextFee = GetTollFee(date, vehicle);
            int tempFee = GetTollFee(intervalStart, vehicle);

            long diffInMillies = date.Ticks - intervalStart.Ticks;
            double minutes = diffInMillies / 10000000 / 60;

            if (minutes < 60.0)
            {
                if (totalFee > 0) totalFee -= tempFee;
                if (nextFee >= tempFee) tempFee = nextFee;
                totalFee += tempFee;
            }
            else
            {
                totalFee += nextFee;
                intervalStart = date;
            }
        }

            if (totalFee > 60) totalFee = 60;

        return totalFee;
    }


    public static int GetTollFee(DateTime date, IVehicle vehicle)
    {
        if (IsTollFreeDate(date) || vehicle.IsTollFree) return 0;

        int hour = date.Hour;
        int minute = date.Minute;

        if (hour == 6 && minute >= 0 && minute <= 29) return 8;
        else if (hour == 6 && minute >= 30 && minute <= 59) return 13;
        else if (hour == 7 ) return 18;
        else if (hour == 8 && minute >= 0 && minute <= 29) return 13;
        else if (hour == 8 && minute >= 30 && minute <= 59) return 8;
        else if (hour > 8 && hour <= 14 ) return 8;
        else if (hour == 15 && minute >= 0 && minute <= 29) return 13;
        else if (hour == 15 && minute >= 30 && minute <= 59) return 18;
        else if (hour == 16 ) return 18;
        else if (hour == 17 ) return 13;
        else if (hour == 18 && minute >= 0 && minute <= 29) return 8;
        else return 0;
    }

    private static bool IsTollFreeDate(DateTime date)
    {
        // Free on weekends
        if (date.DayOfWeek == DayOfWeek.Saturday || date.DayOfWeek == DayOfWeek.Sunday) 
            return true;

        // Use NuGet package Nager.Date to check if it is a holiday!
        if (DateSystem.IsPublicHoliday(date, CountryCode.SE))
            return true;
            
        // Otehrwise return false
        return false;
    }

}